### PR TITLE
Handle type conversion errors in FilterByValueBlock

### PR DESCRIPTION
**Pull request by russellb**
97d3fcd filterblock: Don't break when filtering on unexpected data
111711f Add unit tests for sdg.filterblock

commit 97d3fcd6513320edf4195ba62940af3eb79ffd8f
Author: Russell Bryant <rbryant@redhat.com>
Date:   Wed Jul 3 19:53:23 2024 -0400

    filterblock: Don't break when filtering on unexpected data
    
    The previous code in this block did filtering assuming that all
    samples had a value that was correct for the type. For example, when
    filtering on an integer value, it assumed every row had a valid
    integer, where it may instead have garbage.
    
    This change introduces a new helper, _convert_dtype(), which properly
    handles this condition. When the conversion fails on a `ValueError`
    exception, it treats it as `None` instead of allowing the exception to
    be raised up to the caller.
    
    The fix was authored by Shiv in PR #72. I only pulled it out into a
    standalone commit.
    
    Signed-off-by: Russell Bryant <rbryant@redhat.com>
    Co-authored-by: shiv <shivchander.s30@gmail.com>

commit 111711fa419b53e2f31b63bcb91f829eacd55548
Author: Russell Bryant <rbryant@redhat.com>
Date:   Wed Jul 3 19:56:45 2024 -0400

    Add unit tests for sdg.filterblock
    
    This new file introduces some unit tests for the filterblock module.
    A previous commit introduced a bug fix to ensure that unexpected data
    does not cause this block to break. These tests exercise those fixes.
    The tests introduced here will run automatically in CI and help ensure
    that the problem does not reoccur.
    
    Signed-off-by: Russell Bryant <rbryant@redhat.com>

### CODE REVIEW

**Comment by russellb** [comment]
I posted this as part of a discussion thread here: https://github.com/instructlab/sdg/pull/72#discussion_r1664920262

File: tests/test_filterblock.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+            convert_dtype=int,
+        )
+        self.dataset = Dataset.from_dict(
+            {"age": ["25", "30", "35", "forty", "45"]},
+            features=Features({"age": Value("string")}),
+        )
+
+    def test_convert_dtype_success(self):
+        sample = {"age": "25"}
+        converted_sample = self.block._convert_dtype(sample)
```
Comment (markmc) [inline tests/test_filterblock.py:28]: FWIW I think I would have been inclined to not test the internal `_convert_dtype()` but instead exercise it just through Block.generate()

**Comment by markmc** [comment]
Great, very clear. Well done on adding tests. lgtm

File: tests/test_filterblock.py (hunk 1 above)
Comment (russellb) [reply tests/test_filterblock.py:28]: Makes sense. The last case at the end does that. I could just remove these tests of the internal thing. It’d be less code but we still have the same test coverage.

Comment (russellb) [reply tests/test_filterblock.py]: I removed the tests of the internal method as suggested, since they don't actually add additional test coverage

**Comment by russellb** [comment]
CI is passing, it was approved previously as part of #72, and @markmc reviewed this version that split it out and added tests, so I'm going to merge

**Pull request merged**